### ISSUE TITLE
FieldException: Attempt to create an instance of field og_membership_request on bundle og_membership_type_default that already has an instance of that field. in field_create_instance()

### ISSUE THREAD

**Issue by argiepiano** [backdrop-contrib/og#34]
## Bug description
If you completely uninstall OG and then re-install, you'll get this fatal error:
`FieldException: Attempt to create an instance of field og_membership_request on bundle og_membership_type_default that already has an instance of that field. in field_create_instance()`

## Steps to reproduce
1. Install OG
2. Completely uninstall it
3. Install it again and watch the error appear

## Additional information
The field  instances for`og_membership_request` and two others are  supposedly deleted in `og.uninstall()`. But this is actually not happening. 

I believe the correct way to remove those fields is by invoking:
`field_attach_delete_bundle()`

I haven't really tested this with OG, but I did with another module that also added and attached fields during installation and that seemed to work. This may have to do with some core issues.

ALSO, in my experience, you may not be able to use `field_info_instances()` to get all the bundles where these fields are attached. From what I've seen, in Backdrop this invocation comes too late (again, probably a bug in core), and doesn't return anything at this point. Again, haven't done much testing with this, but I think I had to actually do a db_select to get the bundles.

**Comment by argiepiano** [comment]
See https://github.com/backdrop/backdrop-issues/issues/5469 for a similar problem

**Comment by laryn** [comment]
I have confirmed this, even after all the test-related fixes.

**Comment by argiepiano** [comment]
This is a core bug that has been reported in several shapes and forms but has not been fixed, I believe.

**Comment by argiepiano** [comment]
This issue also exists in D7. See https://www.drupal.org/project/og/issues/1511894

**Comment by argiepiano** [comment]
PR #113 submitted. I'll leave here for a few days before merging in case someone else wants to test.

Explanation: there were two problems:
1. Field `og_membership_request` not being removed at all (field info, field instance and data). This happens also in the D7 version. The entity to which it's attached, `og_membership`, has been deleted before reaching `og_uninstall()`, and therefore the removal of the field needed some special treatment. The current approach in `og_uninstall()` skipped the removal since the entity bundle did not exist anymore and was not retrievable through `field_info_instances()`. The solution was to use `field_delete_instance()` 
2. Other OG fields attached to core entities like nodes (e.g. `group_group` and `og_group_ref`) were marked as deleted (their info and instance config files), but they, and the field storage table were not removed. Running cron can take care of removing everything. So, to solve this without waiting until cron ran, I used `field_purge_batch` at the end of the uninstall hook, as recommended in some D7 posts. This removes any items marked for deletion.

Tested thoroughly - working fine.
